### PR TITLE
Use upstream tile, fuse, and distribution for codegen driver approaches.

### PR DESCRIPTION
**Pull request by hanhanW**

### CODE REVIEW

**Comment by hanhanW** [comment]
This depends on https://github.com/google/iree/pull/8613

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:220]: I dont know if that is going to be enough. The `TileAndDistributeToWorkgroups` uses https://github.com/google/iree/blob/main/iree/compiler/Codegen/Common/DestructiveUpdateUtils.h to convert from a loop form where values are yielded to a form where you use `flow.dispatch.tensor.load/strore`. Without that things might not connect as easily with IREE?
Its great if it works already, but I doubt it does.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp
Comment (hanhanW) [reply iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:220]: It's fine because CSE/Canonicalizer will remove those after IREEComprehensiveBufferize. The IREEComprehensiveBufferize itself would create scf.for loop which returns memref even we don't have rewriteLinalgDestructiveUpdate. The CSE patterns are kicked in and address similar issue. I'll double check if it applies to most of cases. It works for fill + matmul + generic op.

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:220]: Thats awesome actually!

File: iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/Passes.cpp:115]: Could we relax this a bit. If unset the inner loops are treated as tile size set to 0. I dont think we need to explicitly enforce this?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Overall looks great! Just one question.

File: iree/test/e2e/regression/lowering_config.mlir
Comment (MaheshRavishankar) [inline iree/test/e2e/regression/lowering_config.mlir:2]: If we relax the verifier, this change is unnecessary.

File: iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp
Comment (hanhanW) [reply iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:220]: Good news! I enabled the flag locally and found that we pass all the checks. There are no stack allocation in IREE's tests.

I think we still have some stack allocation in tensorflow integration tests like https://github.com/google/iree/issues/8592

Actually, I don't really think we can remove the alloca ops in this case. All the results of generic ops are needed during the computation, and we drop one of final results at the end. We may want to revisit if the allocation can be deleted before switching to upstream tile+fuse+distribute.

File: iree/test/e2e/regression/lowering_config.mlir
Comment (hanhanW) [reply iree/test/e2e/regression/lowering_config.mlir:2]: This is needed for some reasons. It's really tricky. Say that we have fill + matmul ops.

If only two tiling sizes are set, the strategy would use `fill` op as the anchor op. This stops fusion between fill + matmul.

If the three tiling sizes are set, the `matmul` op would be the anchor op. Then the fusion works.

Comment (MaheshRavishankar) [reply iree/test/e2e/regression/lowering_config.mlir:2]: That is strange, I'd rather fix that.... Couple of questions that can help diagnose
1) Which operation has the attribute? Do both fill and matmul have the attribute? If thats the case, thats the bug (potentially here https://github.com/google/iree/blob/d487b5854bf38c8b9466899c583a081ad3d59bd8/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L879)
2) If not the bug is probably here (https://github.com/google/iree/blob/d487b5854bf38c8b9466899c583a081ad3d59bd8/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp#L54).

Could we look into that before we go ahead with this?

Comment (hanhanW) [reply iree/test/e2e/regression/lowering_config.mlir:2]: 1) Only matmul op carries the attribute -- which is expected.

One of the issues is how root op is determined. I checked that it is bottom up approach when anchor-op is not specified. In this case, linalg.matmul is picked in the first place. However the pattern is not kicked in. (I also verified it with manually specifying the anchor op to be linalg.matmul)

https://github.com/llvm/llvm-project/blob/5c116d50e42f93ef4fa9a9719121378ec6b99b69/mlir/lib/Dialect/Linalg/Transforms/LinalgStrategyPasses.cpp#L68-L72

The issue turns into why tile and fuse does not work when the anchor-op is linalg.matmul. This is the debug message:

```
Processing operation : 'linalg.matmul'(0xccdfff0) {
  * Pattern mlir::linalg::LinalgTileAndFuseTensorOpsPattern : 'linalg.matmul -> ()' {
Trying to match "mlir::linalg::LinalgTileAndFuseTensorOpsPattern"
    ** Failure : expect #tile sizes >= #loops
"mlir::linalg::LinalgTileAndFuseTensorOpsPattern" result 0
  } -> failure : pattern failed to match
} -> failure : pattern failed to match
```

There is a check that the tiling sizes should be equal to or greater than the number of loops. I think it is a relaxable condition because it is acceptable during tiling. We append zeros when the number of tiling sizes is less than the number of loops. I'm not sure why we have this condition upstream. @gyist should weigh in since he has more context about fusion.

Comment (hanhanW) [reply iree/test/e2e/regression/lowering_config.mlir:2]: Here is the check: https://github.com/llvm/llvm-project/blob/5c116d50e42f93ef4fa9a9719121378ec6b99b69/mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp#L561-L571

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/test/e2e/regression/lowering_config.mlir
Comment (MaheshRavishankar) [reply iree/test/e2e/regression/lowering_config.mlir:2]: Actually that might be a bug on our side. We should pick up the configuration from matmul, but the first linalg op in the bottum up tranversal should be the root.

**Comment by hanhanW** [comment]
We have correctness issue because we don't apply rewriteLinalgDestructiveUpdates. This is needed after IREE distribution. I'm going to create a new pass which calls the method, and add the pass right after the distribution.

IR after distribution: https://gist.githubusercontent.com/hanhanW/eacd5c7cbdad43afc8d172217cda33b9/raw

```mlir
%6 = scf.for %arg0 = %workgroup_id_z to %c2 step %workgroup_count_z iter_args(%arg1 = %cst) -> (tensor<2x2x2x2x2x2x2xf32>) {
  ...
}
flow.dispatch.tensor.store %6, %1, offsets = [0, 0, 0, 0, 0, 0, 0], sizes = [2, 2, 2, 2, 2, 2, 2], strides = [1, 1, 1, 1, 1, 1, 1] : tensor<2x2x2x2x2x2x2xf32> -> !flow.dispatch.tensor<writeonly:2x2x2x2x2x2x2xf32>
```

It should be replaced with

```mlir
scf.for ... {
   %res = linalg.generic ...
   flow.dispatch.tensor.store %res ...
}
```

**Comment by hanhanW** [comment]
@MaheshRavishankar please take another look, thank you!

**Review by MaheshRavishankar — APPROVED** [review]
Now that it is separate we can test the tile and distribute and rewrite separately. I'm not holding you to it. I did find a bug in the rewrite update, so I can move my test into this file. Thanks!

**Comment by ScottTodd** [comment]
`iree/test/e2e/matmul/e2e_matmul_direct_f32_large_split_k_dylib-llvm-aot_dylib` is newly failing after merging this
sample logs: https://source.cloud.google.com/results/invocations/cf1a5bf4-c896-420d-9f01-3a8998aa6c58/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fmain/log

**Comment by MaheshRavishankar** [comment]
Fun! Mid-air collision. @hanhanW if its not a simple fix, might be worth reverting for now.

**Comment by hanhanW** [comment]
I'm not sure if it could be a simple fix or not actually. Let's revert it for now. I'm fine with resubmitting the PR.

**Comment by ThomasRaoux** [comment]
ouch, I wonder if there is a bug on my side, let me know if you need help debugging

**Pull request merged**
